src/stan/io/deserializer.hpp

```diff
@@ -0,0 +1,1978 @@
+   * @tparam Sizes A parameter pack of integral types.
+   * @param lp The reference to the variable holding the log
+   * @param sizes Pack of integrals to use to construct the return's type.
+   * probability to increment.
+   * @return The next unit_vector of the specified size.
+   * @throw std::invalid_argument if k is zero
+   */
+  template <typename Ret, bool Jacobian, typename LP, typename... Sizes,
+            require_not_std_vector_t<Ret>* = nullptr>
+  auto read_unit_vector(LP& lp, Sizes... sizes) {
+    using stan::math::unit_vector_constrain;
+    if (Jacobian) {
+      return unit_vector_constrain(this->read<Ret>(sizes...), lp);
+    } else {
+      return unit_vector_constrain(this->read<Ret>(sizes...));
+    }
+  }
+
+  /**
+   * Return the next unit_vector of the specified size (using one fewer
+   * unconstrained scalars), incrementing the specified reference with the
+   * log absolute Jacobian determinant.
+   *
+   * <p>See <code>stan::math::unit_vector_constrain(Eigen::Matrix,T&)</code>.
+   *
+   * @tparam Ret The type to return.
+   * @tparam Jacobian Whether to increment the log of the absolute Jacobian
+   * determinant of the transform.
+   * @tparam LP Type of log probability.
+   * @tparam Sizes A parameter pack of integral types.
+   * @param lp The reference to the variable holding the log
+   * @param vecsize The size of the return vector.
+   * @param sizes Pack of integrals to use to construct the return's type.
+   * probability to increment.
+   * @return The next unit_vector of the specified size.
+   * @throw std::invalid_argument if k is zero
+   */
+  template <typename Ret, bool Jacobian, typename LP, typename... Sizes,
+            require_std_vector_t<Ret>* = nullptr>
+  auto read_unit_vector(LP& lp, const size_t vecsize, Sizes... sizes) {
+    std::decay_t<Ret> ret;
+    ret.reserve(vecsize);
+    for (size_t i = 0; i < vecsize; ++i) {
+      ret.emplace_back(
+          this->read_unit_vector<value_type_t<Ret>, Jacobian>(lp, sizes...));
+    }
+    return ret;
+  }
+
+  /**
+   * Return a simplex of the specified size made up of the
+   * next scalars.
+   *
+   * <p>See <code>stan::math::check_simplex</code>.
+   *
+   * @tparam Ret The type to return.
+   * @param k Size of returned simplex.
+   * @return Simplex read from the specified size number of scalars.
+   * @throw std::runtime_error if the k values is not a simplex.
+   * @throw std::invalid_argument if k is zero
+   */
+  template <typename Ret, require_vector_t<Ret>* = nullptr,
+            require_not_std_vector_t<Ret>* = nullptr>
+  auto read_simplex(size_t k) {
+    if (unlikely(k == 0)) {
+      []() STAN_COLD_PATH {
+        std::string msg = "deserializer: simplex vectors cannot be size 0.";
+        throw std::invalid_argument(msg);
+      }();
+    }
+    auto ret = read<Ret>(k);
+    using stan::math::check_simplex;
+    using stan::math::value_of;
+    check_simplex("deserializer", "Simplex", value_of(ret));
```

Comment:
`check_simplex` checks for zero sizes ([here](https://github.com/stan-dev/math/blob/develop/stan/math/prim/err/check_simplex.hpp)) so the check above the read is not necessary